### ISSUE TITLE
Switching loss type mid-training + tools repository

### ISSUE THREAD

**Issue by artdgn** [lyst/lightfm#258]
Hi,
First, thanks for creating and maintaining this great package!

Second, I have a couple of questions:

1) I noticed that on my data WARP trains much much faster than BPR, but for some cases, BPR's final accuracy is better. Also WARP sometimes overfits (validation accuracy goes down with more epochs), and with BPR I haven't observed this. 
In your opinion, does it make sense to train with WARP first and than switch to BPR for "fine tuning"? If so, what would be the best way to do this? (I see that there are various state variables for the loss and weight updates, wonder how to do it right). I've also noticed that (on my data), WARP only works well with adadelta, and BPR only with adagrad, would changing learning schedules work as well?

2) I have several implicit feedback signals (several types of interactions e.g. views and purchases), from what I've read so far from the issues, the way to correctly use them is by specifying them in sample_weight matrix. But in other places I read that it won't matter much (because the models are primarity implicit feedback ones). What is your suggestion in terms of learning from different types of interactions (all implicit, but of different significance)? Is sample_weight the right way, or do I need to create several models?

3) I have quite a lot of utility and tool code already: sampling recommendations, similarities, early stop fitting, hyperparameter tuning (via skopt), dataframes as inputs and outputs (hiding away the sparse matrices), feature matrix construction (out of an additional dataframe via sklearn-pandas), evaluation utilities.. What would be a good way to contribute this code? I read somewhere that you don't want to introduce external package dependencies, so would you prefer that as a separate repository, e.g. "lightfm-tools" or something of that sort (because there are quite a lot of pydata packages in my code)?

**Comment by maciejkula** [comment]
1. One interesting characteristic of WARP is that it is effectively playing against itself: it is preferentially using examples that it scores highly as negative examples. Up to a certain point, this allows it to express user preferences better; beyond that point, it is self-defeating, as it is selecting the examples that it should recommend and treating them as negatives. The extent to which that happens is governed by the `max_sampled` [hyperparameter](http://lyst.github.io/lightfm/docs/lightfm.html#lightfm.LightFM).

I suspect that this is what happening in your case. Instead of switching to BPR, can you try decreasing that hyperparameter? Otherwise, I don't see any obvious problems with switching losses. It's interesting that you find different results with adagrad and adadelta, I haven't observed that.

2. `sample_weight` is the right way of using this in the LightFM model. If you are optimizing for some weighted objective function, changing the weights here will probably help in achieving your goal.

3. I'm open to a submodule along the lines of `lightfm.contrib` which would contain high-quality, but not strictly speaking supported functionality (say, with optional dependencies). I'd be happy to discuss your additions (but I have to warn I can be quite picky!). While I personally am not a huge fan of `pandas`, I know it's quite popular, so maybe input from pandas dataframes would be a good first step?

**Comment by artdgn** [comment]
1. Thanks, tried tuning that parameter as well. So far it seems that increasing this parameter a little bit increases accuracy. So I guess I'll leave that at that and will explore switching losses, or decreasing that parameter mid-training later.

2. Thanks for the confirmation.

3. Great. I'll create a generic LighFM-tools repo from what I'm working on, this way some of it might be useful to people even before all of it is of sufficient quality to be included in `contrib` to the main package. Will keep you updated.

**Comment by RAbraham** [comment]
@artdgn Looking forward to your contributions!

**Comment by RAbraham** [comment]
@maciejkula Just curious, what do you use in place of `pandas`?

**Comment by maciejkula** [comment]
@RAbraham I use plain Python objects and numpy arrays.

Pandas is a useful tool, but I prefer not using it.

**Comment by artdgn** [comment]
Hi @maciejkula, Had some time to put together an open repo with the tools I'm using, with LightFM's at it's core. 
https://github.com/DomainGroupOSS/ml-recsys-tools/

Hopefully it might be useful to someone else and reduce the learning curve and time to results.

Naturally since I've been working on it alone and for the specific uses of the project I'm working on, it's by no means as generic or bulletproof, and is still a work in progress. 

Would be grateful for any feedback or ideas to make it more useful. I intend to add examples (there's [just one](https://github.com/DomainGroupOSS/ml-recsys-tools/blob/master/examples/basic_example.py) now), documentation and tests.

**Comment by maciejkula** [comment]
@artdgn thanks for this, I will have a look. Which parts do you think would be useful for most users?

I also noticed you have a [helper for building interaction matrices](https://github.com/DomainGroupOSS/ml-recsys-tools/blob/master/ml_recsys_tools/data_handlers/interaction_handlers_base.py#L192). I've been working on something similar to include in LightFM itself, and would love to have your comments (https://github.com/lyst/lightfm/pull/281)!

**Comment by maciejkula** [comment]
How do you deal with categorical features in pandas? Load it in as strings, change type to categorical, then one-hot encode?

**Comment by artdgn** [comment]
> How do you deal with categorical features in pandas? Load it in as strings, change type to categorical, then one-hot encode?

I load as (or convert to) strings and one-hot encode ([here](https://github.com/DomainGroupOSS/ml-recsys-tools/blob/df866704b1ab3ad05dbe40ebd4b2d92d071f8eab/ml_recsys_tools/data_handlers/interactions_with_features.py#L154))
The more tricky part is to one-hot encode numerical features, which is done with a custom transformer (same place). As I found that using the numerical features as is (scaled) seems to not work as well (maybe because the relationships are rarely linear? Don't know if that makes sense).

>  think doing the name -> idx mappings correctly and consistently is the tricky thing, because we're dealing with two objects: features and interactions (as you say). As far as constructing sparse matrices goes, I'm pretty sure my approach is about as efficient as you can get without dropping down to compiled languages (much more efficient that scipy's DOK and LIL matrices).
Can you point me to a place where you do the dataframe -> sparse matrices translation?

The matrix construction is quick once you have the indexes, it's the conversion that I found needed to be parallelized ([this](https://github.com/DomainGroupOSS/ml-recsys-tools/blob/df866704b1ab3ad05dbe40ebd4b2d92d071f8eab/ml_recsys_tools/data_handlers/interaction_handlers_base.py#L237) is how I'm doing it for now). I'm handling the conversion using sklearn LabelEncoders (they allow to transform and inverse_transform), and I'm converting and creating the matrix in the same place. I've noticed you're mapping using a dict but are converting IDs to idx in a loop, which is where I'd guess I'd expect to have a bottleneck.

The alignment of the feature matrix is done on creation of the features dataframe, by using the relevant encoder - [here](https://github.com/DomainGroupOSS/ml-recsys-tools/blob/df866704b1ab3ad05dbe40ebd4b2d92d071f8eab/ml_recsys_tools/data_handlers/interactions_with_features.py#L100).

**Comment by majd1239** [comment]
> Can you point me to a place where you do the dataframe -> sparse matrices translation?

Was just reading this and thought my input could help. I recently used lightfm to train a 17GB dataset with 420M rows for 10M unique  users and 20k unique items.
The raw data was a pandas dataframe of `user_id, item_id, score`. Using the Dataset module to build the interaction matrices was really slow. Even on a sample of 1M, it took a very long time.

My initial approach was converting to `category` dtypes and then use the cat.codes as mappings, but with large dataframes even changing the dtype takes a long time. What I did was simply construct a manual mapping for indices and ids. 
Eg:
```
unique_user_ids = df.user_id.unique( )
num_user_ids = len(unique_user_ids)
user_ids_mapping =  dict( zip(unique_user_ids , range(num_user_ids) )
# below method is slow, what I did was split the df in chunks and applied the mapping in parallel then # concated the chunks back.
df.user_id = df.user_id.apply(lambda id: user_ids_mapping[id] )
```
And then same for item_ids

Then to get the coo matrices:
```
coo_matrix( (df.score , (df.user_id , df.item_id)), dtype=np.float32)
```

The whole mapping, transforming and constructing the matrices took under 3 mins for the whole dataset. Keeping in mind the dataset has 420 Million rows. 
This can be easily extended to accommodate user_features and item_features. Point is using numpy/pandas  vectorization to construct the interactions matrices is much faster than the current setup of native python looping/appending

**Comment by SimonCW** [comment]
I’m closing this issue because it has been inactive for a long time. If you still encounter the problem, please open a new issue.

Thank you!
